src/fun/transform/expand_generated.rs

```diff
@@ -1,24 +1,107 @@
+use std::collections::{BTreeSet, HashMap, HashSet};
+
 use crate::{
-  fun::{Book, Term},
+  fun::{Book, Name, Term},
   maybe_grow,
 };
 
-/// Dereferences any generated definitions in the term.
+/// Dereferences any non recursive generated definitions in the term.
 /// Used after readback.
 impl Term {
-  pub fn expand_generated(&mut self, book: &Book) {
+  pub fn expand_generated(&mut self, book: &Book, recursive_defs: &RecursiveDefs) {
     maybe_grow(|| {
       if let Term::Ref { nam } = &*self {
-        if nam.is_generated() {
+        if nam.is_generated() && !recursive_defs.contains(nam) {
           *self = book.defs.get(nam).unwrap().rule().body.clone();
         }
       }
 
-      // Note: this assumes that there will never be a loop of generated functions.
-      // This is true right now, but not necessarily in the future.
       for child in self.children_mut() {
-        child.expand_generated(book);
+        child.expand_generated(book, recursive_defs);
       }
     })
   }
 }
+
+type DepGraph = HashMap<Name, HashSet<Name>>;
+type Cycles = Vec<Vec<Name>>;
+type RecursiveDefs = BTreeSet<Name>;
+
+pub fn find_recursive_defs(book: &Book) -> RecursiveDefs {
+  let mut cycle_map = BTreeSet::new();
+  let deps = book_def_deps(book);
+  let cycles = cycles(&deps);
+
+  for cycle in cycles {
+    for name in cycle {
+      cycle_map.insert(name);
+    }
+  }
+
+  cycle_map
+}
+
+/// Find all cycles in the dependency graph.
+fn cycles(deps: &DepGraph) -> Cycles {
+  let mut cycles = vec![];
+  let mut stack = vec![];
+  let mut visited = HashSet::new();
+  for nam in deps.keys() {
+    if !visited.contains(nam) {
+      find_cycles(deps, nam, &mut visited, &mut stack, &mut cycles);
+    }
+  }
+  cycles
+}
+
+fn find_cycles(
+  deps: &DepGraph,
+  nam: &Name,
+  visited: &mut HashSet<Name>,
+  stack: &mut Vec<Name>,
+  cycles: &mut Cycles,
+) {
+  maybe_grow(|| {
+    // Check if the current ref is already in the stack, which indicates a cycle.
+    if let Some(cycle_start) = stack.iter().position(|n| n == nam) {
+      // If found, add the cycle to the cycles vector.
+      cycles.push(stack[cycle_start..].to_vec());
+      return;
+    }
+    // If the ref has not been visited yet, mark it as visited.
+    if visited.insert(nam.clone()) {
+      // Add the current ref to the stack to keep track of the path.
+      stack.push(nam.clone());
```

Comment:
Since this function is already stack based, you could make it a while loop and not recursive right?